### ISSUE TITLE
[YSQL] Wrong results in a query combining cross join lateral and CTEs

### ISSUE THREAD

**Issue by soupi** [yugabyte/yugabyte-db#20297]
Jira Link: [DB-9263](https://yugabyte.atlassian.net/browse/DB-9263)
### Description

Hello. I've run into an issue on `yugabytedb/yugabyte:2.19.3.0-b140` with the following query:

```sql
SELECT
  row_to_json("result") AS "result"
FROM
  json_to_recordset(
    cast(
      '[{"%variable_order": 1, "search": "%Quest%"}, {"%variable_order": 2, "search": "%Garage%"}]' AS json
    )
  ) AS "%variables_table"("search" varchar, "%variable_order" int)
  CROSS JOIN LATERAL (
    WITH my_album AS (
      SELECT
        *
      FROM
        public."Album"
      WHERE
        "Title" LIKE cast("%variables_table"."search" AS varchar)
    )
    SELECT
      "%variables_table"."%variable_order",
      "my_album"."Title"
    FROM
      "my_album"
  ) AS "result"
ORDER BY
  "%variables_table"."%variable_order" ASC
```

results:

```
                     result                      
-------------------------------------------------
 {"%variable_order":1,"Title":"Jota Quest-1995"}
 {"%variable_order":2,"Title":"Jota Quest-1995"}
(2 rows)
```

This is on a standard chinook database.

The idea behind this query is to run a query for each set of variables in the `%variables_table` table. However, it seems like we are running the same set of variables every time.

For comparison, this is the expected result that postgres returns:

```
                        result                        
------------------------------------------------------
 {"%variable_order":1,"Title":"Jota Quest-1995"}
 {"%variable_order":2,"Title":"Garage Inc. (Disc 1)"}
 {"%variable_order":2,"Title":"Garage Inc. (Disc 2)"}
(3 rows)
```

So it seems like there's some issue with the combination of CTEs and cross join lateral.

An interesting thing to note though, is that if I return a `%variables_table` column from the CTE query, I get the expected results:

```sql
SELECT
  row_to_json("result") AS "result"
FROM
  json_to_recordset(
    cast(
      '[{"%variable_order": 1, "search": "%Quest%"}, {"%variable_order": 2, "search": "%Garage%"}]' AS json
    )
  ) AS "%variables_table"("search" varchar, "%variable_order" int)
  CROSS JOIN LATERAL (
    WITH my_album AS (
      SELECT
        "%variables_table"."%variable_order", -- !!!
        *
      FROM
        public."Album"
      WHERE
        "Title" LIKE cast("%variables_table"."search" AS varchar)
    )
    SELECT
      "%variables_table"."%variable_order",
      "my_album"."Title"
    FROM
      "my_album"
  ) AS "result"
ORDER BY
  "%variables_table"."%variable_order" ASC
```

results from yugabye:

```
                        result                        
------------------------------------------------------
 {"%variable_order":1,"Title":"Jota Quest-1995"}
 {"%variable_order":2,"Title":"Garage Inc. (Disc 2)"}
 {"%variable_order":2,"Title":"Garage Inc. (Disc 1)"}
(3 rows)
```

Thanks.

### Issue Type

kind/bug

### Warning: Please confirm that this issue does not contain any sensitive information

- [X] I confirm this issue does not contain any sensitive information.

[DB-9263]: https://yugabyte.atlassian.net/browse/DB-9263?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by ddorian** [comment]
Hi @soupi 

Can you share the schema/content of your `Album` table so we can reproduce?

**Comment by sushantrmishra** [comment]
This issue is duplicate of https://github.com/yugabyte/yugabyte-db/issues/19694 and https://github.com/yugabyte/yugabyte-db/issues/20316

**Comment by tanujnay112** [comment]
Thank you for finding this @soupi. This regression had been fixed on master as per #19694 but that fix had not been backported to earlier releases which we're currently working on as quickly as we can. For now, you'll be able to work around this regression by setting the GUC flag `yb_enable_expression_pushdown` to false either as a session variable as:
```
SET yb_enable_expression_pushdown = false;`
```

or as a per database config:

```
ALTER DATABASE <target_db_name> SET yb_enable_expression_pushdown = false;
```

**Comment by m-iancu** [comment]
Duplicate of #20316.
Fixed as part of https://github.com/yugabyte/yugabyte-db/issues/19694, dedicated test added in [b64b536](https://github.com/yugabyte/yugabyte-db/commit/b64b5365f3372bfafc7b824ef04b62ac456d52b0).
